### PR TITLE
AgentCoder-like Example for Coding Solutions

### PR DESCRIPTION
**Pull request by dillonalaird**
First PR for adding in AgentCoder. This is just an example of how to get an agent to code, but some of the abstractions here, like tools_v2, I want to use for other coding examples.

Typing issues: It looks like Windows doesn't support a lot of the resource or signal methods/constants, so I have a few `@typing.no_type_check`'s in there to ignore a lot of the typing issues. If you add too many `# type: ignore` then Linux/Mac will complain about unnecessary `# type: ignore`'s

### CODE REVIEW

File: vision_agent/agent/agent_coder.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    of the code written by the coder agent. If the code fails, we pass it back to the
+    coder agent to fix the bug, if it succeeds we pass it to a visual tester agent, which
+    is an LMM model like GPT4V, to visually inspect the output and make sure it looks
+    good."""
+
+    def __init__(
+        self,
+        coder_agent: Optional[LLM] = None,
+        tester_agent: Optional[LLM] = None,
+        visual_tester_agent: Optional[LMM] = None,
```
Comment (shankar-landing-ai) [inline vision_agent/agent/agent_coder.py:101]: So the visual_tester_agent acts like reflection ?

File: vision_agent/agent/agent_coder.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        verbose: bool = False,
+    ) -> None:
+        self.coder_agent = (
+            OpenAILLM(temperature=0.1) if coder_agent is None else coder_agent
```
Comment (shankar-landing-ai) [inline vision_agent/agent/agent_coder.py:105]: Do we need any specialized coding LLM for these tasks  or for now we use GPT4V

File: vision_agent/agent/agent_coder.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        question = input[0]["content"]
+        if image:
+            question += f" Input file path: {os.path.abspath(image)}"
+
+        code = ""
+        feedback = ""
+        for _ in range(self.max_turns):
+            code = write_program(question, feedback, self.coder_agent)
+            _LOGGER.info(f"code:\n{code}")
+            debug = write_debug(question, code, feedback, self.tester_agent)
```
Comment (shankar-landing-ai) [inline vision_agent/agent/agent_coder.py:142]: What happens for tasks where we don't have visualizations ? Like Image to text models. It just executes the code ?

File: vision_agent/agent/execution.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+                with time_limit(timeout):
+                    # WARNING
+                    # This program exists to execute untrusted model-generated code. Although
+                    # it is highly unlikely that model-generated code will do something overtly
+                    # malicious in response to this test suite, model-generated code may act
+                    # destructively due to a lack of model capability or alignment.
+                    # Users are strongly encouraged to sandbox this evaluation suite so that it
+                    # does not perform destructive actions on their host or network.
+                    # Once you have read this disclaimer and taken appropriate precautions,
+                    # uncomment the following line and proceed at your own risk:
```
Comment (shankar-landing-ai) [inline vision_agent/agent/execution.py:65]: Just making sure there is nothing to uncomment here ?

**Review by shankar-landing-ai — COMMENTED** [review]
Left few comments, more for my understanding. Overall looks good

File: vision_agent/agent/agent_coder.py (hunk 1 above)
Comment (dillonalaird) [reply vision_agent/agent/agent_coder.py:101]: Yeah this is very similar to vision agent, where the reflection agent is an LMM that looks at the visualized output.

File: vision_agent/agent/agent_coder.py (hunk 2 above)
Comment (dillonalaird) [reply vision_agent/agent/agent_coder.py:105]: No, just GPT4-turbo here is good (it's currently top in coding on LMSYS)

File: vision_agent/agent/execution.py (hunk 4 above)
Comment (dillonalaird) [reply vision_agent/agent/execution.py:65]: Yeah this is from the original code base, I left a comment to link the original code at the top of the file. I will replace this file pretty soon with the notebook code execution

**Review by shankar-landing-ai — APPROVED** [review]
LGTM

**Pull request merged**